trie-db/src/lookup.rs

```diff
@@ -655,4 +668,219 @@ where
 		}
 		Ok(None)
 	}
+
+	/// Look up the merkle value (hash) of the node that is the closest descendant for the provided
+	/// key.
+	///
+	/// When the provided key leads to a node, then the merkle value of that node
+	/// is returned. However, if the key does not lead to a node, then the merkle value
+	/// of the closest descendant is returned. `None` if no such descendant exists.
+	fn inner_lookup_first_descendent(
+		mut self,
+		nibble_key: NibbleSlice,
+		full_key: &[u8],
+	) -> Result<Option<TrieHash<L>>, TrieHash<L>, CError<L>> {
+		let mut partial = nibble_key;
+		let mut hash = self.hash;
+		let mut key_nibbles = 0;
+
+		// Consume the cache similar to `Self::look_up`.
+		let mut cache = self.cache.take();
+
+		// this loop iterates through non-inline nodes.
+		for depth in 0.. {
+			// Ensure the owned node reference lives long enough.
+			// Value is never read, but the reference is.
+			let mut _owned_node = NodeOwned::Empty;
+
+			// The binary encoded data of the node fetched from the database.
+			//
+			// Populated by `get_owned_node` to avoid one extra allocation by not
+			// calling `NodeOwned::to_encoded` when computing the hash of inlined nodes.
+			let mut node_data = Vec::new();
+
+			// Get the owned node representation from the database.
+			let mut get_owned_node = |depth: i32| {
+				let data = match self.db.get(&hash, nibble_key.mid(key_nibbles).left()) {
+					Some(value) => value,
+					None =>
+						return Err(Box::new(match depth {
+							0 => TrieError::InvalidStateRoot(hash),
+							_ => TrieError::IncompleteDatabase(hash),
+						})),
+				};
+
+				let decoded = match L::Codec::decode(&data[..]) {
+					Ok(node) => node,
+					Err(e) => return Err(Box::new(TrieError::DecoderError(hash, e))),
+				};
+
+				let owned = decoded.to_owned_node::<L>()?;
+				node_data = data;
+				Ok(owned)
+			};
+
+			let mut node = if let Some(cache) = &mut cache {
+				let node = cache.get_or_insert_node(hash, &mut || get_owned_node(depth))?;
+
+				self.record(|| TrieAccess::NodeOwned { hash, node_owned: node });
+
+				node
+			} else {
+				_owned_node = get_owned_node(depth)?;
+
+				self.record(|| TrieAccess::EncodedNode {
+					hash,
+					encoded_node: node_data.as_slice().into(),
+				});
+
+				&_owned_node
+			};
```

Comment:
Now we always need to create an owned node, which is a little bit shitty. But fine for now I think.